File: android/app/src/main/java/com/calcplus/calculator/core/lock/PasscodeRules.kt
package com.calcplus.calculator.core.lock

/**
 * Shared validation for first-run setup, the change flow and the switch flow.
 * Face-independent by construction: it sees opaque token IDs only.
 */
object PasscodeRules {
    const val MIN_TOKENS = 4
    const val MAX_TOKENS = TokenRecorder.MAX_TOKENS

    fun isValidLength(tokens: List<String>): Boolean =
        tokens.size in MIN_TOKENS..MAX_TOKENS

    /** Soft warning only, never a block: a single repeated token (e.g. 7777). */
    fun isTrivial(tokens: List<String>): Boolean =
        tokens.isNotEmpty() && tokens.toSet().size == 1
}

File: ios/SafeBox/Lock/PasscodeRules.swift
import Foundation

/// Shared validation for first-run setup and the Settings change flows, so the
/// rules cannot drift apart. Face-agnostic: the unit is a token ID.
enum PasscodeRules {
    static let minTokens = 4
    static let maxTokens = 32

    static func isValidLength(_ tokens: [String]) -> Bool {
        tokens.count >= minTokens && tokens.count <= maxTokens
    }

    /// Soft warning only, never a block: a single repeated token (e.g. 7777).
    static func isTrivial(_ tokens: [String]) -> Bool {
        guard !tokens.isEmpty else { return false }
        return Set(tokens).count == 1
    }
}
